lib/alloc.h

```diff
@@ -20,6 +20,58 @@
 #include "defines.h"
 
 
+#define CALLOC(n, type)        ((type *) calloc(n, sizeof(type)))
+#define XCALLOC(n, type)       ((type *) xcalloc(n, sizeof(type)))
+#define MALLOCARRAY(n, type)   ((type *) mallocarray(n, sizeof(type)))
+#define XMALLOCARRAY(n, type)  ((type *) xmallocarray(n, sizeof(type)))
+
+#define ALLOCA(type)           ALLOCARRAY(1, type)
+#define MALLOC(type)           MALLOCARRAY(1, type)
+#define XMALLOC(type)          XMALLOCARRAY(1, type)
+#define REALLOC(ptr, type)     REALLOCARRAY(ptr, 1, type)
+#define REALLOCF(ptr, type)    REALLOCARRAYF(ptr, 1, type)
+
+#define REALLOCARRAY(ptr, n, type)                                            \
+({                                                                            \
+	__auto_type  p_ = (ptr);                                              \
+                                                                              \
+	static_assert(__builtin_types_compatible_p(typeof(p_), type *), "");  \
```

Comment:
static_assert() appears to be a c++ thing?

Reply:
It's in C11:

`static_assert(3)`:
```
static_assert(3)           Library Functions Manual           static_assert(3)

NAME
       static_assert, _Static_assert - fail compilation if assertion is false

LIBRARY
       Standard C library (libc)

SYNOPSIS
       #include <assert.h>

       void static_assert(scalar constant-expression, const char *msg);

       /* Since C23: */
       void static_assert(scalar constant-expression);

DESCRIPTION
       This  macro is similar to assert(3), but it works at compile time, gen‐
       erating a compilation error (with an optional message) when  the  input
       is false (i.e., compares equal to zero).

       If the input is nonzero, no code is emitted.

       msg must be a string literal.  Since C23, this argument is optional.

       There's  a keyword, _Static_assert(), that behaves identically, and can
       be used without including <assert.h>.

RETURN VALUE
       No value is returned.

VERSIONS
       In C11, the second argument (msg) was mandatory; since C23, it  can  be
       omitted.

STANDARDS
       C11 and later.

EXAMPLES
       static_assert()  can't  be  used  in  some  places, like for example at
       global scope.  For that, a macro must_be() can be written in  terms  of
       static_assert().   The following program uses the macro to get the size
       of an array safely.

           #include <assert.h>
           #include <stddef.h>
           #include <stdint.h>
           #include <stdio.h>
           #include <stdlib.h>
           #include <string.h>

           /*
            * This macro behaves like static_assert(), failing to
            * compile if its argument is not true.  However, it always
            * returns 0, which allows using it everywhere an expression
            * can be used.
            */
           #define must_be(e)                                        \
           (                                                         \
               0 * (int) sizeof(                                     \
                   struct {                                          \
                       static_assert(e);                             \
                       int  ISO_C_forbids_a_struct_with_no_members;  \
                   }                                                 \
               )                                                     \
           )

           #define is_same_type(a, b)  \
               __builtin_types_compatible_p(typeof(a), typeof(b))

           #define is_array(arr)       (!is_same_type((arr), &*(arr)))
           #define must_be_array(arr)  must_be(is_array(arr))

           #define sizeof_array(arr)   (sizeof(arr) + must_be_array(arr))
           #define nitems(arr)         (sizeof((arr)) / sizeof((arr)[0]) \
                                        + must_be_array(arr))

           int     foo[10];
           int8_t  bar[sizeof_array(foo)];

           int
           main(void)
           {
               for (size_t i = 0; i < nitems(foo); i++) {
                   foo[i] = i;
               }

               memcpy(bar, foo, sizeof_array(bar));

               for (size_t i = 0; i < nitems(bar); i++) {
                   printf("%d,", bar[i]);
               }

               exit(EXIT_SUCCESS);
           }

SEE ALSO
       assert(3)

Linux man-pages (unreleased)        (date)                    static_assert(3)
```
<https://manpages.debian.org/testing/manpages-dev/static_assert.3.en.html>

Reply:
Hm, my 22.04 system does not have that manpage, making me very unhappy.

However, it seems to be compiling fine here so we'll roll with it.

(I probably shouldn't admit that whenever I see C11 I think "but that's so recent"... )

Reply:
> Hm, my 22.04 system does not have that manpage,

I wrote it on:

4a96ff90d - Fri, 2 Dec 2022 16:11:46 +0100 (3 months ago)
            _Static_assert.3, static_assert.3: Add documentation for static_assert() - Alejandro Colomar

so it's to be expected :)

> making me very unhappy.

You can always get them from upstream ;)

(or use Debian Sid, which picks them pretty fast.  I guess you maybe need some drivers for a laptop that only Ubuntu has though?)

> 
> However, it seems to be compiling fine here so we'll roll with it.
> 
> (I probably shouldn't admit that whenever I see C11 I think "but that's so recent"... )

I think the opposite when I see C89 (it fells like "nobody really took the time to update this in the last 24 years?" :p).
I learnt C with C11 already out there (it was around that year that I started with C, +/-1), so I feels natural to me.